code/obj/machinery/spaceheater.dm

```diff
@@ -8,216 +8,176 @@ TYPEINFO(/obj/machinery/space_heater)
 	icon_state = "sheater0"
 	name = "space HVAC"
 	desc = "Made by Space Amish using traditional space techniques, this space heater is guaranteed not to set the station on fire."
-	var/emagged = 0
+	var/emagged = FALSE
 	var/obj/item/cell/cell
-	var/on = 0
-	var/heating = 0
-	var/open = 0
-	var/set_temperature = 50		// in celcius, add T0C for kelvin
-	var/heating_power = 40000
-	var/cooling_power = -30000
+	var/on = FALSE
+	var/heating = FALSE // If its cooling down (false) or heating up (true) the current atmosphere
+	var/set_temperature = 293 // Room temperature
+	var/heating_power = 100000
+	var/cooling_power = -100000
 	deconstruct_flags = DECON_WRENCH | DECON_WELDER
-	flags = FPRINT
+	flags = FPRINT | TGUI_INTERACTIVE
 
 
 	New()
 		..()
-		cell = new(src)
-		cell.charge = 1000
-		cell.maxcharge = 1000
+		src.cell = new(src)
+		src.cell.charge = 1000
+		src.cell.maxcharge = 1000
 		UpdateIcon()
 		return
 
+	ui_interact(mob/user, datum/tgui/ui)
+		ui = tgui_process.try_update_ui(user, src, ui)
+		if(!ui)
+			ui = new(user, src, "space_heater", src.name)
+			ui.open()
+
+	ui_data(mob/user)
+		. = list()
+		.["on"] = src.on
+		.["max"] = src.emagged ? 1000 : 328
+		.["min"] = src.emagged ? 0 : 258
+		.["heating"] = src.heating
+		.["emagged"] = src.emagged
+		.["set_temperature"] = src.set_temperature
+		.["cell"] = src.cell
+		if (src.cell != null) // Cant get the cell variables if it doesnt exist
+			.["cell_name"] = src.cell.name
+			.["cell_charge"] = src.cell.percent()
+
 	update_icon()
-		if (on)
-			if(heating)
+		if (src.on)
+			if(src.heating)
 				icon_state = "sheaterH"
 			else
 				icon_state = "sheaterC"
 		else
 			icon_state = "sheater0"
-		if(open)
-			icon_state = "sheater-open"
 		return
 
 	emag_act(var/mob/user, var/obj/item/card/emag/E)
 		if (!src.emagged)
-			if (open)
-				if (user)
-					user.show_text("You short out the temperature limiter circuit in the [src].", "blue")
-				src.emagged = 1
-				src.desc = "Made by Space Amish using traditional space techniques, this heater is guaranteed to set the station on fire."
-				return 1
-			else
-				if (user)
-					user.show_text("The internal circuitry must be exposed!", "red")
-				return 0
+			if (user)
+				user.show_text("You short out the temperature limiter circuit in the [src].", "blue")
+			src.emagged = TRUE
+			src.desc = "Made by Space Amish using traditional space techniques, this heater is guaranteed to set the station on fire."
```

Comment:
I dont understand, do you mean replacing src.desc="string" with src.get_desc()="string"? I dont understand the benefit of that or how it relates to user.show_text() as well since that is only supposed to be shown once to the player.

Reply:
get_desc is a function on basically everything and overriding it is usually the preferred method on doing changing descriptions
see `/obj/item/sword/discount` or `/obj/item/plate/pizza_box`

can't see how it relates to the rest

Reply:
Ah, so it works practically the same as examine(), I guess im just gonna replace examine() with get_desc() and add an if statement to change the description in there.